CodeEdit/Features/Settings/Pages/TextEditingSettings/Models/TextEditingSettings.swift

```diff
@@ -239,6 +254,55 @@ extension SettingsData {
                 }
             }
         }
+
+        struct InvisibleCharactersConfig: Equatable, Hashable, Codable {
+            static var `default`: InvisibleCharactersConfig = {
+                InvisibleCharactersConfig(
+                    enabled: false,
+                    showSpaces: true,
+                    showTabs: true,
+                    showLineEndings: true
+                )
+            }()
+
+            var enabled: Bool
+
+            var showSpaces: Bool
+            var showTabs: Bool
+            var showLineEndings: Bool
+
+            var spaceReplacement: String = "·"
+            var tabReplacement: String = "→"
+
+            // Controlled by `showLineEndings`
+            var carriageReturnReplacement: String = "↵"
+            var lineFeedReplacement: String = "¬"
+            var paragraphSeparatorReplacement: String = "¶"
+            var lineSeparatorReplacement: String = "⏎"
+        }
+
+        struct WarningCharacters: Equatable, Hashable, Codable {
+            static let `default`: WarningCharacters = WarningCharacters(enabled: true, characters: [
```

Comment:
Looks good, maybe this isa good addition too https://www.compart.com/en/unicode/U+037E

Reply:
Oh that's a good one, I'll add it.